### PR TITLE
Enhance OpenBinary Reader

### PR DESCRIPTION
**Pull request by pedromxavier**
This PR aims to:
- Provide meaningful error messages when parsing graf headers.
- Add support for non-positive stage indices.

### CODE REVIEW

File: src/OpenBinary/reader.jl [hunk 1]
```diff
@@ -133,18 +125,17 @@ function PSRI.open(
         _first_stage = read(ioh, Int32) #month or week
         first_year = read(ioh, Int32) # year
         #
-        unit_buffer = Vector{Cchar}(undef, 7)
-        read!(ioh, unit_buffer)
-        unit_str = strip(join(unit_buffer))
+        unit_str = strip(join(Char.(read!(ioh, Vector{Cchar}(undef, 7)))))
```
Comment (guilhermebodin) [inline src/OpenBinary/reader.jl:128]: tricky to read this one. IMO this should be either separated or written with pipes to be more clear `read!(ioh, Vector{Cchar}(undef, 7)) .|> Char |> join |> strip`

File: src/OpenBinary/reader.jl [hunk 2]
```diff
@@ -156,137 +147,157 @@ function PSRI.open(
         _first_stage = read(ioh, Int32) #month or week
         first_year = read(ioh, Int32) # year
         #
-        unit_buffer = Vector{Cchar}(undef, 7)
-        read!(ioh, unit_buffer)
-        unit_str = strip(join(Char.(unit_buffer)))
+        unit_str = strip(join(Char.(read!(ioh, Vector{Cchar}(undef, 7)))))
```
Comment (guilhermebodin) [inline src/OpenBinary/reader.jl:150]: same here

File: src/OpenBinary/reader.jl [hunk 3]
```diff
@@ -156,137 +147,157 @@ function PSRI.open(
...
             number_blocks = zeros(Int32, stage_total + 1 - first_relative_stage)
             offsets = zeros(Int32, 1 + stage_total + 1 - first_relative_stage)
-            for i in first_relative_stage:stage_total+1
-                offset = read(ioh, Int32)
-                offsets[i-first_relative_stage+1] = offset
-                if i > first_relative_stage
-                    number_blocks[i-first_relative_stage] = offset - offsets[i-first_relative_stage]
+            
+            for i in first_relative_stage:(stage_total+1)
+                let offset = read(ioh, Int32)
```
Comment (guilhermebodin) [inline src/OpenBinary/reader.jl:176]: Why do you need a let statement here? Maybe consider renaming these offsets to some more informative name such as `header_offset`, `stage_offset`, etc.

File: src/OpenBinary/reader.jl [hunk 4]
```diff
@@ -156,137 +147,157 @@ function PSRI.open(
...
+    end
 
-    @assert stage_total > 0
-    @assert first_relative_stage == 1 # todo improve this
-    @assert stage_total - first_relative_stage >= 0
+    # TODO: improve this
+    # if first_relative_stage != 1
+    #     @info   first_relative_stage
+    #     @assert false
+    # end
```
Comment (guilhermebodin) [inline src/OpenBinary/reader.jl:225]: Remove comment

File: src/OpenBinary/reader.jl [hunk 5]
```diff
@@ -156,137 +147,157 @@ function PSRI.open(
...
 
     if variable_by_series == 0
         # previous versions of the file simply leave the number of scenarios
         # as the number of scenarios in the study
         scenario_total = 1
     end
+
     if !(0 <= variable_by_series <= 1)
-        println(BAD * "variable_by_series = $variable_by_series, expected 0 or 1")
+        @warn("'variable_by_series' must be either 0 or 1, not $(variable_by_series)")
```
Comment (guilhermebodin) [inline src/OpenBinary/reader.jl:238]: Maybe this should be an error

File: src/OpenBinary/reader.jl [hunk 6]
```diff
@@ -156,137 +147,157 @@ function PSRI.open(
...
+        error("'scenario_total' must be a positive integer, not $(scenario_total)")
     end
-    @assert scenario_total > 0
-    scenario_exist = variable_by_series == 1
+
+    scenario_exist = (variable_by_series == 1)
 
     if !(0 <= variable_by_block <= 3)
-        println(BAD * "variable_by_block = $variable_by_block, expected 0, 1, 2 or 3")
+        @warn("'variable_by_block' must be 0, 1, 2 or 3, not $(variable_by_block)")
```
Comment (guilhermebodin) [inline src/OpenBinary/reader.jl:248]: Maybe it should also be an error.

File: src/OpenBinary/reader.jl [hunk 7]
```diff
@@ -156,137 +147,157 @@ function PSRI.open(
...
+    
+    if !(block_total > 0)
+        error("'block_total' must be a positive integer, not $(block_total)")
+    end
+
+    block_exist = (variable_by_block > 0)
 
     if !(0 <= variable_by_hour <= 1)
-        println(BAD * "variable_by_hour = $variable_by_hour, expected 0 or 1")
+        @warn("'variable_by_hour' must be 0, 1, 2 or 3, not $(variable_by_hour)")
```
Comment (guilhermebodin) [inline src/OpenBinary/reader.jl:262]: Also an error?

File: src/OpenBinary/reader.jl [hunk 8]
```diff
@@ -156,137 +147,157 @@ function PSRI.open(
...
     end
-    hours_exist = variable_by_hour == 1
+
+    hours_exist = (variable_by_hour == 1)
 
     if total_agents == 0
-        println("total_agents == 0, nothing to do")
-        # todo
+        @warn("No agents were provided") # todo
```
Comment (guilhermebodin) [inline src/OpenBinary/reader.jl:268]: This could be more of an info.

File: src/OpenBinary/reader.jl [hunk 9]
```diff
@@ -156,137 +147,157 @@ function PSRI.open(
...
-            _agent = strip(agent)
-            ret = findfirst(x -> x == _agent, agent_names)
-            if ret === nothing
-                # println("agent $(_agent) not found in file")
-                error("agent $(_agent) not found in file")
+
+        index = sizehint!(Int[], length(header))
+
+        for agent in strip.(header)
+            ret = findfirst(x -> x == agent, agent_names)
```
Comment (guilhermebodin) [inline src/OpenBinary/reader.jl:281]: `ret = findfirst(isequal(agent), agent_names)`

File: src/OpenBinary/reader.jl [hunk 10]
```diff
@@ -156,137 +147,157 @@ function PSRI.open(
...
-        index = Int[]
-        sizehint!(index, length(header))
-        for agent in header
-            _agent = strip(agent)
-            ret = findfirst(x -> x == _agent, agent_names)
-            if ret === nothing
-                # println("agent $(_agent) not found in file")
-                error("agent $(_agent) not found in file")
+
+        index = sizehint!(Int[], length(header))
```
Comment (guilhermebodin) [inline src/OpenBinary/reader.jl:278]: Why is this not `Vector{Int}(undef, lenght(header)`? It should find everyone, if it doesn`t It will error anyway

File: src/OpenBinary/reader.jl [hunk 11]
```diff
@@ -156,137 +147,157 @@ function PSRI.open(
...
+            if isnothing(ret)
+                error("Agent '$(agent)' not found in file")
             end
+            
             push!(index, ret)
         end
+
         if !allow_empty && isempty(index)
-            error("no agents found" * ifelse(length(header) == 0, ", empty header inserted. If you do not want to pass a header use: use_header = false option.", "."))
+            if length(header) == 0
```
Comment (guilhermebodin) [inline src/OpenBinary/reader.jl:291]: `if isempty(header)` is a bit easier to read for me, personal opinion.

File: src/OpenBinary/reader.jl [hunk 12]
```diff
@@ -355,30 +354,26 @@ function PSRI.open(
...
-        if x.is_open
-            Base.close(x.io)
+    finalizer(ret) do ptr
+        if ptr.is_open
+            Base.close(ptr.io)
         end
-        x
+
+        return ptr
     end
```
Comment (guilhermebodin) [inline src/OpenBinary/reader.jl:367]: Maybe read more about it here https://docs.julialang.org/en/v1/base/base/#Base.finalizer. Does it need an `@async` call

File: src/OpenBinary/reader.jl [hunk 13]
```diff
@@ -453,29 +448,123 @@ function PSRI.goto(graf::Reader, t::Integer, s::Integer = 1, b::Integer = 1)
...
+        end
+    elseif stage_type == PSRI.STAGE_DAY
+        24
+    elseif stage_type == PSRI.STAGE_YEAR
+        8760
+    else
+        error("Stage type $(stage_type) is not currently supported")
+    end
+
+    return floor(Int, block_total / hour_total)
```
Comment (guilhermebodin) [inline src/OpenBinary/reader.jl:521]: Not that it makes a huge difference but... https://docs.julialang.org/en/v1/manual/performance-tips/#Tweaks 
try to use `fld`

File: src/OpenBinary/writer.jl [hunk 14]
```diff
@@ -231,25 +238,25 @@ function PSRI.open(
     end
 
     if single_binary
-        write(ioh, Int32(0)) #!!!
+        write(ioh, Int32(0))               # !!! ???
```
Comment (guilhermebodin) [inline src/OpenBinary/writer.jl:241]: Indeed important to understand why is this

**Review by guilhermebodin — CHANGES_REQUESTED** [review]
Looks good but I have some minor requests. Also it would be nice to add some tests on truncated files and negative stage ones too.

File: src/OpenBinary/reader.jl [hunk 15]
```diff
@@ -133,18 +125,17 @@ function PSRI.open(
...
+        unit_str = strip(join(Char.(read!(ioh, Vector{Cchar}(undef, 7)))))
         #
         name_length = read(ioh, Int32)
 
         ##
         variable_by_hour = 0
         number_blocks = Int32[]
     else
-        skip_store(ioh, 4, skips, store_skips)
-        skip_store(ioh, 4, skips, store_skips)
```
Comment (joaquimg) [inline src/OpenBinary/reader.jl:147]: You gave up on this?
Whats the reason?

File: src/OpenBinary/reader.jl [hunk 16]
```diff
@@ -156,137 +147,157 @@ function PSRI.open(
...
+    if !single_binary
         close(ioh)
     end
 
-    BAD = "Bad header: "
+    if !(stage_total > 0)
+        error("'stage_total' must be a positive integer, not '$(stage_total)'")
+    end
 
-    @assert stage_total > 0
```
Comment (joaquimg) [inline src/OpenBinary/reader.jl:238]: We should at least check it this is larger or equal zero

File: src/OpenBinary/reader.jl [hunk 17]
```diff
@@ -156,137 +147,157 @@ function PSRI.open(
...
-    @assert first_relative_stage == 1 # todo improve this
-    @assert stage_total - first_relative_stage >= 0
+    # TODO: improve this
+    # if first_relative_stage != 1
+    #     @info   first_relative_stage
+    #     @assert false
+    # end
+
+    if !(first_relative_stage <= stage_total)
```
Comment (joaquimg) [inline src/OpenBinary/reader.jl:227]: Should we check if the first relative is negative or zero?

File: src/OpenBinary/reader.jl (hunk 10 above)
Comment (joaquimg) [reply src/OpenBinary/reader.jl:278]: Not exactly... it might be useful to allow for an incomplete header

Comment (joaquimg) [reply src/OpenBinary/reader.jl:278]: We would still have to fix other things for it to work properly

File: src/OpenBinary/reader.jl [hunk 18]
```diff
@@ -453,29 +448,123 @@ function PSRI.goto(graf::Reader, t::Integer, s::Integer = 1, b::Integer = 1)
...
-    if PSRI.is_hourly(graf)
+function _get_relative_offset(ptr::Reader)
+    return 4 * ptr.agents_total * ptr.scenario_total * (ptr.first_relative_stage - 1)
+end
+
+function _get_relative_offset(::Any)
+    return 0
+end
+
+function _get_position(ptr::Any, t::Integer, s::Integer, b::Integer)
```
Comment (joaquimg) [inline src/OpenBinary/reader.jl:459]: Dont type as "Any", just leave is blank

File: src/OpenBinary/reader.jl [hunk 19]
```diff
@@ -453,29 +448,123 @@ function PSRI.goto(graf::Reader, t::Integer, s::Integer = 1, b::Integer = 1)
     return nothing
 end
 
-function _get_position(graf, t::Integer, s::Integer, b::Integer)
-    if PSRI.is_hourly(graf)
+function _get_relative_offset(ptr::Reader)
+    return 4 * ptr.agents_total * ptr.scenario_total * (ptr.first_relative_stage - 1)
+end
+
+function _get_relative_offset(::Any)
```
Comment (joaquimg) [inline src/OpenBinary/reader.jl:455]: This is necessary?

File: src/OpenBinary/reader.jl (hunk 5 above)
Comment (pedromxavier) [reply src/OpenBinary/reader.jl:238]: I just turned `print`s into `@warn`s in this commit. We can discuss that.

File: src/OpenBinary/reader.jl (hunk 12 above)
Comment (pedromxavier) [reply src/OpenBinary/reader.jl:367]: According to the docs, it needs. `Base.close` triggers a `flush` call, that is, an I/O operation. Thus, `@async` is required.
I added it.

File: src/OpenBinary/reader.jl (hunk 13 above)
Comment (pedromxavier) [reply src/OpenBinary/reader.jl:521]: @joaquimg @guilhermebodin Taking a closer look at this function I ask myself: Why keeping `ptr.is_open` instead of calling `Base.isopen` on `ptr.io`?

Comment (pedromxavier) [reply src/OpenBinary/reader.jl:521]: This function is called once per load. Conceptually, `div` is better suited since operands are integers.

File: src/OpenBinary/reader.jl (hunk 16 above)
Comment (pedromxavier) [reply src/OpenBinary/reader.jl:238]: The error message that was added already indicates which value was given

File: src/OpenBinary/reader.jl (hunk 17 above)
Comment (pedromxavier) [reply src/OpenBinary/reader.jl:227]: The implementation accounts for this by subtracting this as an offset where necessary. Any relative stage (negative, zero, positive) should work.

File: src/OpenBinary/reader.jl (hunk 19 above)
Comment (pedromxavier) [reply src/OpenBinary/reader.jl:455]: Yes. The `Writer` also uses `_get_position`.

Comment (pedromxavier) [reply src/OpenBinary/reader.jl:455]: And, yes, it is not aware of the relative offset yet. If you read a 0-index graf and then attempt to write it, it will get 1-indexed.

Part II - Enhance OpenBinary Reader is coming soon.

File: src/OpenBinary/reader.jl (hunk 15 above)
Comment (pedromxavier) [reply src/OpenBinary/reader.jl:147]: It was important when I was investigating that binary data alignment issue last year. Now that it is stable we don't need it anymore.

File: src/OpenBinary/reader.jl (hunk 12 above)
Comment (guilhermebodin) [reply src/OpenBinary/reader.jl]: Let's leave like this and change in another PR, could you create an issue about it?

**Comment by guilhermebodin** [comment]
@joaquimg @pedromxavier, is it good to go? Is there something missing?

**Comment by pedromxavier** [comment]
@guilhermebodin new tests are still missing. We've already passed the old ones, but they did not cover non-positive indices anyway
